### PR TITLE
Add binaris server(less) option

### PR DESCRIPTION
**Pull request by bergundy**
More binaris tests required.
After merge need to use `serverCommon.ts` in `koa` backend and remove `koa-json-error`.

### CODE REVIEW

File: templates/ts/fn-binaris.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        {{/serverContext}}
+        {{^serverContext}}
+        return await fn({{#parameters}}args.{{{name}}}{{^last}}, {{/last}}{{/parameters}});
+        {{/serverContext}}
+      } catch (error) {
+        const { stack, ...rest } = error;
+        const err = stackTraceInError ? { stack: stack.toString(), ...rest } : rest;
+        {{#throws}}
+        if (error instanceof {{{.}}}) {
+          return new ctx.HTTPResponse({
```
Comment (michaeladda) [inline templates/ts/fn-binaris.ts:70]: this can be DRY with `HTTPJsonResponse(statusCode, object)` or something similar

Comment (bergundy) [reply templates/ts/fn-binaris.ts:70]: I'm going to refactor this into a helper to share between `binaris` and `koa` servers.

File: README.md [hunk 2]
```diff
@@ -249,6 +249,40 @@ When null is the only return type on a method, as in `returns: null`, it will co
...
+      const client = new ExampleClient(url, {
+        headers: {
+          'X-Binaris-Api-Key': process.env.BINARIS_API_KEY,
+        },
+      });
+      console.log('1 + 2 =', await client.add(1, 2));
+    }
+    main();
+    ```
+1. Run your test with `node test.js`, should print out `1 + 2 = 3` (That is if I can do math ;) ).
```
Comment (vogre) [inline README.md:284]: This only works when environment variables defined:

```shell
BINARIS_ACCOUNT_ID=$(bn show accountId) BINARIS_API_KEY=$(bn show apiKey) node test.js
```

File: templates/ts/fn-binaris.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        {{/serverOnlyContext}}
+        const context = { ...clientContext, ...serverOnlyContext };
+        {{#serverContext}}
+        return await fn(context{{#parameters}}, args.{{{name}}}{{/parameters}});
+        {{/serverContext}}
+        {{^serverContext}}
+        return await fn({{#parameters}}args.{{{name}}}{{^last}}, {{/last}}{{/parameters}});
+        {{/serverContext}}
+      } catch (error) {
+        const { stack, ...rest } = error;
```
Comment (vogre) [inline templates/ts/fn-binaris.ts]: This doesn't work for `message` (and possibly other non-own/non-enumerable properties, so `err.message` is undefined)

File: templates/ts/fn-binaris.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+              'Content-Type': 'application/json',
+            },
+            body: JSON.stringify({
+              name: 'ValidationError',
+              message: err.message,
+              errors: err.errors,
+            }),
+          });
+        }
+        return new ctx.HTTPResponse({
```
Comment (vogre) [inline templates/ts/fn-binaris.ts:99]: `stackTraceInError` is not respected for the `InternalServerError`

File: templates/ts/fn-binaris.ts (hunk 1 above)
Comment (vogre) [reply templates/ts/fn-binaris.ts:70]: Note that returning an error response means the error is only returned to clients and not logged in logs, so it makes it harder for debug. We probably should `console.error` the error by default here.

File: src/test/test_binaris.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+${this.tester}`);
+
+    await spawn('npm', [
+      'install',
+      'chai',
+      'chai-as-promised',
+      '@types/chai',
+      '@types/chai-as-promised',
+    ], {
+      cwd: this.dir,
```
Comment (vogre) [inline src/test/test_binaris.ts]: DRY for `{ cwd, stdio }`?

**Review by vogre — CHANGES_REQUESTED** [review]
Thanks!

Left a few comments about logs, error messages, and running locally.

File: README.md (hunk 2 above)
Comment (bergundy) [reply README.md:284]: 3 steps back mention that you have to export your binaris creds.

File: templates/ts/fn-binaris.ts (hunk 3 above)
Comment (bergundy) [reply templates/ts/fn-binaris.ts]: Good catch

File: templates/ts/fn-binaris.ts (hunk 4 above)
Comment (bergundy) [reply templates/ts/fn-binaris.ts:99]: Why not?

**Review by bergundy — COMMENTED** [review]
Thanks!

File: README.md (hunk 2 above)
Comment (vogre) [reply README.md:284]: The steps are really confusing because they are all numbered 1!

File: README.md [hunk 6]
```diff
@@ -249,6 +249,40 @@ When null is the only return type on a method, as in `returns: null`, it will co
...
+    const { ExampleClient } = require('./client');
+
+    async function main() {
+      const url = `https://run.binaris.com/v2/run/${process.env.BINARIS_ACCOUNT_ID}`;
+      const client = new ExampleClient(url, {
+        headers: {
+          'X-Binaris-Api-Key': process.env.BINARIS_API_KEY,
+        },
+      });
+      console.log('1 + 2 =', await client.add(1, 2));
```
Comment (vogre) [inline README.md:280]: I suggest `try-catch` here for nicer display of possible errors (like forgot to export keys).

File: templates/ts/fn-binaris.ts (hunk 4 above)
Comment (vogre) [reply templates/ts/fn-binaris.ts:99]: Hmm... You're right, it's only the message I was missing.

Note that the errors are somewhat different when the error is in Binaris itself (e.g. you make an error defining the wrapper): you get `{ error: 'blah', stack: 'blahblah' }`.

**Review by vogre — APPROVED** [review]
Thanks!

I don't think there are any blocking issues, but what do you think about logging the errors?

File: templates/ts/fn-binaris.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            },
+            body: JSON.stringify({
+              name: 'ValidationError',
+              message,
+              errors: error.errors,
+            }),
+          });
+        }
+        // tslint:disable-next-line:no-console
+        console.error(error);
```
Comment (bergundy) [inline templates/ts/fn-binaris.ts:83]: @vogre we you looking for this?

**Pull request merged**